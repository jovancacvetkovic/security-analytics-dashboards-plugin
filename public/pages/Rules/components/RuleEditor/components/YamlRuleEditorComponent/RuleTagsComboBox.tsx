/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import { EuiFormRow, EuiText, EuiComboBox, EuiComboBoxOptionOption } from '@elastic/eui';

export interface RuleTagsComboBoxProps {
  onCreateOption: (
    searchValue: string,
    options: EuiComboBoxOptionOption<string>[]
  ) => boolean | void;
  onBlur: any;
  onChange: ((options: EuiComboBoxOptionOption<string>[]) => void) | undefined;
  selectedOptions: EuiComboBoxOptionOption<string>[];
}

const STARTS_WITH = 'attack.';

const isValid = (value: string) => {
  if (value === '') return true;
  return value.startsWith(STARTS_WITH) && value.length > STARTS_WITH.length;
};

export const RuleTagsComboBox: React.FC<RuleTagsComboBoxProps> = ({
  onCreateOption,
  onBlur,
  onChange,
  selectedOptions,
}) => {
  const [isCurrentlyTypingValueInvalid, setIsCurrentlyTypingValueInvalid] = useState(false);

  const onSearchChange = (searchValue: string) => {
    setIsCurrentlyTypingValueInvalid(!isValid(searchValue));
  };

  return (
    <>
      <EuiFormRow
        label={
          <EuiText size={'s'}>
            <strong>Tags </strong>
            <i>- optional</i>
          </EuiText>
        }
        isInvalid={isCurrentlyTypingValueInvalid}
        error={isCurrentlyTypingValueInvalid ? 'Invalid tag' : ''}
        helpText={`Tags must start with '${STARTS_WITH}'`}
      >
        <EuiComboBox
          noSuggestions
          onSearchChange={onSearchChange}
          placeholder="Create tags"
          onChange={onChange}
          onCreateOption={(searchValue, options) =>
            isValid(searchValue) && onCreateOption(searchValue, options)
          }
          onBlur={onBlur}
          data-test-subj={'rule_tags_dropdown'}
          selectedOptions={selectedOptions}
          isInvalid={isCurrentlyTypingValueInvalid}
        />
      </EuiFormRow>
    </>
  );
};
