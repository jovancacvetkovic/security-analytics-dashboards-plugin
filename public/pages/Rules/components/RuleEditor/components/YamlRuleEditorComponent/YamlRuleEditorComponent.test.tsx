/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from '@testing-library/react';
import { YamlRuleEditorComponent } from './YamlRuleEditorComponent';

describe('<YamlRuleEditorComponent /> spec', () => {
  it('renders the component', () => {
    const { container } = render(
      <YamlRuleEditorComponent
        change={() => {}}
        rule={{
          id: '25b9c01c-350d-4b95-bed1-836d04a4f324',
          category: 'windows',
          title: 'Testing rule',
          description: 'Testing Description',
          status: 'experimental',
          author: 'Bhabesh Raj',
          references: [
            {
              value: 'https://securelist.com/operation-tunnelsnake-and-moriya-rootkit/101831',
            },
          ],
          tags: [
            {
              value: 'attack.persistence',
            },
            {
              value: 'attack.privilege_escalation',
            },
            {
              value: 'attack.t1543.003',
            },
          ],
          log_source: '',
          detection:
            'selection:\n  Provider_Name: Service Control Manager\n  EventID: 7045\n  ServiceName: ZzNetSvc\ncondition: selection\n',
          level: 'high',
          false_positives: [
            {
              value: 'Unknown',
            },
          ],
        }}
        isInvalid={false}
      >
        <div>Testing YamlRuleEditor</div>
      </YamlRuleEditorComponent>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component - invalid state', () => {
    const { container } = render(
      <YamlRuleEditorComponent
        change={() => {}}
        rule={{
          id: '25b9c01c-350d-4b95-bed1-836d04a4f324',
          category: 'windows',
          title: 'Testing rule',
          description: 'Testing Description',
          status: 'experimental',
          author: 'Bhabesh Raj',
          references: [
            {
              value: 'https://securelist.com/operation-tunnelsnake-and-moriya-rootkit/101831',
            },
          ],
          tags: [
            {
              value: 'attack.persistence',
            },
            {
              value: 'attack.privilege_escalation',
            },
            {
              value: 'attack.t1543.003',
            },
          ],
          log_source: '',
          detection:
            'selection:\n  Provider_Name: Service Control Manager\n  EventID: 7045\n  ServiceName: ZzNetSvc\ncondition: selection\n',
          level: 'high',
          false_positives: [
            {
              value: 'Unknown',
            },
          ],
        }}
        isInvalid={true}
        errors={['Validation error message']}
      >
        <div>Testing YamlRuleEditor</div>
      </YamlRuleEditorComponent>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
