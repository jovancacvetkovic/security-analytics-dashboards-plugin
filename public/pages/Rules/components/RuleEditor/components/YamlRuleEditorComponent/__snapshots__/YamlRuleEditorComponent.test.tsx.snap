// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<YamlRuleEditorComponent /> spec renders the component - invalid state 1`] = `
<div
  class="euiCallOut euiCallOut--danger"
>
  <div
    class="euiCallOutHeader"
  >
    <span
      class="euiCallOutHeader__title"
    >
      Please address the highlighted errors.
    </span>
  </div>
  <div
    class="euiText euiText--small"
  >
    <div
      class="euiTextColor euiTextColor--default"
    >
      <ul>
        <li>
          Validation error message
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`<YamlRuleEditorComponent /> spec renders the component 1`] = `
<div
  class="euiSpacer euiSpacer--s"
/>
`;
